### ISSUE TITLE
Python backend Cannot Concurrent Load Model

### ISSUE THREAD

**Issue by s402621342** [triton-inference-server/server#5342]
**Description**
Python backend cannot concurrent load model,  model instances start initialize only when the previous instance initialize finish

**Triton Information**
What version of Triton are you using?
r22.12

Are you using the Triton container or did you build it yourself?
using the container nvcr.io/nvidia/tritonserver:22.12-py3

**To Reproduce**
Steps to reproduce the behavior.
A simple way to reproduce 
```
nvidia-docker run -it --rm --shm-size=1g --network=host nvcr.io/nvidia/tritonserver:22.12-py3 bash

```
I have 4 models a,b,c,d and each have 10 instances, model.py is
```
import triton_python_backend_utils as pb_utils
import time 
class TritonPythonModel:

    def initialize(self, args):
        time.sleep(10)

        print('Initialized...')

    def execute(self, requests):

        responses = []

        return responses
```

and I start tritonserver --model-repository=/models --model-load-thread-count=10
the log is 
```
I0210 08:40:32.330140 1720 model_lifecycle.cc:459] loading: a:1
I0210 08:40:32.330184 1720 model_lifecycle.cc:459] loading: b:1
I0210 08:40:32.330204 1720 model_lifecycle.cc:459] loading: c:1
I0210 08:40:32.330224 1720 model_lifecycle.cc:459] loading: d:1
I0210 08:40:37.381442 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: a_0_0 (CPU device 0)
Initialized...
I0210 08:40:47.570451 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: b_0_0 (CPU device 0)
Initialized...
I0210 08:40:57.763593 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: c_0_0 (CPU device 0)
Initialized...
I0210 08:41:07.950528 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: d_0_0 (CPU device 0)
Initialized...
I0210 08:41:18.137234 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: a_0_1 (CPU device 0)
Initialized...
I0210 08:41:28.325216 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: b_0_1 (CPU device 0)
Initialized...
I0210 08:41:38.517562 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: c_0_1 (CPU device 0)
Initialized...
I0210 08:41:48.707506 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: d_0_1 (CPU device 0)
Initialized...
I0210 08:41:58.896498 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: a_0_2 (CPU device 0)
Initialized...
I0210 08:42:09.089506 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: b_0_2 (CPU device 0)
Initialized...
I0210 08:42:19.281562 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: c_0_2 (CPU device 0)
Initialized...
I0210 08:42:29.473462 1720 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: d_0_2 (CPU device 0)
```
You can see clearly, the model instance start initialize only when the previous instance initialize finish

**Expected behavior**
  Model instances should by loaded concurrently with  --model-load-thread-count

**Comment by rmccorm4** [comment]
Hi @s402621342,

I think the Python Backend has some special properties around synchronization compared to other backends, maybe @kthui or @Tabrizian can help explain behavior.

**Comment by kthui** [comment]
There is a global lock acquired inside SharedLibrary when calling any ModelInstanceInitialize function, see the if-statement-block at [L325](https://github.com/triton-inference-server/core/blob/main/src/backend_model_instance.cc#L325). I think there will not be any concurrency before we fully implement parallel model loading. CC @GuanLuo in-case this is not accurate.

**Comment by s402621342** [comment]
Will Python Backend implements parallel model loading is the future release? 
I think in my situation, parallel loading will be helpful when I update the code or need to expand the server pods number

**Comment by kthui** [comment]
> Will Python Backend implements parallel model loading is the future release?

We will implement parallel model loading.

**Comment by oandreeva-nv** [comment]
@kthui Could you please elaborate on parallel model loading for python backend?

**Comment by kthui** [comment]
> it seems that python backend model instances are still loading in serial fashion, is this expected?

Yes, the ***model instances*** of a model are not loaded in parallel.

This issue is about parallel ***model*** loading, please open another issue if you would like to follow up on the feature request. 🙂

**Comment by rmccorm4** [comment]
FYI, Parallel model instance loading for Python and OnnxRuntime backends was added in 23.08 as mentioned here: https://github.com/triton-inference-server/server/issues/5709#issuecomment-1674013011
